src/instrumentation/create-telemetry-middleware.ts

```diff
@@ -0,0 +1,321 @@
+    }
+
+    const traceId = context.emitter?.trace?.id;
+    if (activeTracesMap.has(traceId)) {
+      return;
+    }
+    activeTracesMap.set(traceId, context.instance.constructor.name);
+
+    instrumentationLogger.debug(
+      {
+        source: context.instance.constructor.name,
+        traceId: traceId,
+      },
+      "createTelemetryMiddleware",
+    );
+    const { emitter, runParams, instance } = context;
+    const basePath = createFullPath(emitter.namespace, "");
+
+    let prompt: string | undefined | null = null;
+    if (instance.constructor.name === "BeeAgent") {
+      prompt = (runParams as Parameters<BeeAgent["run"]>)[0].prompt;
+    }
+
+    const spansMap = new Map<string, FrameworkSpan>();
+    const parentIdsMap = new Map<string, number>();
+    const spansToDeleteMap = new Map<string, undefined>();
+
+    let generatedMessage: GeneratedResponse | undefined = undefined;
+    let history: GeneratedResponse[] | undefined = undefined;
+    const groupIterations: string[] = [];
+
+    const idNameManager = new IdNameManager();
+
+    const newTokenEventName: keyof GenerateCallbacks = `newToken`;
+    const partialUpdateEventName: keyof BeeCallbacks = "partialUpdate";
+    const successEventName: keyof GenerateCallbacks = `success`;
+    const finishEventName: keyof GenerateCallbacks = `finish`;
+    const startEventName: keyof GenerateCallbacks = `start`;
+    const errorEventName: keyof GenerateCallbacks = `error`;
+
+    const eventsIterationsMap = new Map<string, Map<string, string>>();
+
+    const startTime = new Date();
+    const serializer = traceSerializer({ ignored_keys: INSTRUMENTATION_IGNORED_KEYS });
+
+    /**
+     * delete all sources related to deleted span
+     * @param param0
+     * @returns
+     */
+    function cleanSpanSources({ spanId }: { spanId: string }) {
+      const parentId = spansMap.get(spanId)?.parent_id;
+      if (!parentId) {
+        return;
+      }
+
+      const spanCount = parentIdsMap.get(parentId);
+      if (!spanCount) {
+        return;
+      }
+
+      if (spanCount > 1) {
+        // increase the span count for the parentId
+        parentIdsMap.set(parentId, spanCount - 1);
+      } else if (spanCount === 1) {
+        // delete the parentId from the map
+        parentIdsMap.delete(parentId);
+        // check the `spansToDelete` if the span should be deleted when it has no children's
+        if (spansToDeleteMap.has(parentId)) {
+          spansMap.delete(parentId);
+          spansToDeleteMap.delete(parentId);
+        }
+      }
+    }
+
+    /**
+     * Create OpenTelemetry spans from collected data
+     */
+    emitter.match(
+      (event) => event.path === `${basePath}.run.${finishEventName}`,
+      async (_, meta) => {
+        try {
+          instrumentationLogger.debug({ path: meta.path, traceId: traceId }, "run finish event");
+
+          if (!prompt && instance.constructor.name === "BeeAgent") {
```

Comment:
I would do `instanceof BeeAgent` instead.

Reply:
I cannot use the `instanceof` in this file for BeeAgent. This is the reason why I use `instance.constructor.name` in more places in this function. Otherwise, I get an error:
```
 FAIL  src/serializer/serializer.test.ts [ src/serializer/serializer.test.ts ]
TypeError: Class extends value undefined is not a constructor or null
 ❯ src/agents/bee/errors.ts:19:36
     17| import { AgentError } from "@/agents/base.js";
     18| 
     19| export class BeeAgentError extends AgentError {}
       |                                    ^
     20| 
 ❯ src/agents/bee/agent.ts:6:31

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  src/agents/bee/runner.spec.ts [ src/agents/bee/runner.spec.ts ]
TypeError: Class extends value undefined is not a constructor or null
 ❯ src/agents/bee/agent.ts:49:31
     47| }
     48| 
     49| export class BeeAgent extends BaseAgent<BeeRunInput, BeeRunOutput, BeeRunOptions> {
       |                               ^
     50|   public readonly emitter = Emitter.root.child<BeeCallbacks>({
     51|     namespace: ["agent", "bee"],
 ❯ src/instrumentation/create-telemetry-middleware.ts:13:32
```

where I run `yarn test:unit`.

> It's probably the **Circular Dependency** import problem.

Reply:
`memory` is available on every `BaseAgent` instance

Reply:
We will soon add new agents.